src/main/java/org/javarosa/xml/TreeElementParser.java

```diff
@@ -55,10 +63,83 @@ public TreeElement parse() throws InvalidStructureException, IOException,
+     * @throws InvalidStructureException
+     * @throws XmlPullParserException
+     * @throws UnfullfilledRequirementsException
+     */
+    public TreeElement parseInternalSecondaryInstance() throws IOException, InvalidReferenceException, InvalidStructureException, XmlPullParserException, UnfullfilledRequirementsException {
+        final int depth = parser.getDepth();
+        //instance node is assumed not to be the first node
+        int foundInstanceIndex = -1;
+        while (parser.getDepth() >= depth) {
+            if(isInstanceNode()){
+                //instance names are in the default namespace
+                String instanceId = parser.getAttributeValue( null, ID_ATTR);
+                if (this.instanceId.equals(instanceId)) {
+                    return new TreeElementParser(parser, foundInstanceIndex, instanceId).parse();
+                } else {
+                    skipSubTree();
+                }
+            }
+        }
+        throw new InvalidReferenceException(String.format("The instance ID %s was not found in the XForm file ", instanceId),instanceId);
+    }
+
+    /**
+     *
+     * Parses the internal instances from an XForm (excluding the main instance)
+     * @return List of #TreeElement representation of the internal
+     * instance nodes.
+     *
+     * @throws IOException There was an error parsing the xml file
+     * @throws InvalidStructureException
+     * @throws XmlPullParserException
+     * @throws UnfullfilledRequirementsException
+     */
+    public List<TreeElement> parseInternalSecondaryInstances() throws InvalidStructureException, XmlPullParserException, UnfullfilledRequirementsException, IOException {
+        /**
+         * This could be implemented using the XPath reference to target specifically
+         * the path of the instance node, but currently, this targets any path where
+         * <instance></instance> node is found
+         */
+        List<TreeElement> internalInstances = new ArrayList<>();
+        final int depth = parser.getDepth();
+        if (depth > 0) {
+            while (parser.getDepth() >= depth) {
+                if(isInstanceNode()){
+                    TreeElement treeElement = new TreeElementParser(parser, 0, "").parse();
+                    if(treeElement.getAttributeValue(null, ID_ATTR) !=null){
+                        String instanceId = treeElement.getAttributeValue(null,ID_ATTR);
+                        if(instanceId != null){
+                            treeElement.setInstanceName(instanceId);
+                            internalInstances.add(treeElement);
+                        }
+                    }
+                }
+            }
+            return internalInstances;
+        } else {
+            throw new InvalidStructureException("Invalid XML File, no detected xml node - Depth is 0");
+        }
+    }
+
+
+    public boolean isInstanceNode() throws IOException, XmlPullParserException {
+        int ret = nextNonWhitespace();
```

Comment:
Would you consider moving this call to the parsing method and passing the token into the boolean method? This method's name implies it's just doing a check but in fact it's consuming a token.